webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultStudentGroupBatchService.java

```diff
@@ -110,27 +109,75 @@ public StudentGroupBatch upload(final User user, final MultipartFile file) {
             return sgbImport;
 
         } catch (final Exception ex) {
+            logger.warn("Upload request failed with exception: {}", file.getOriginalFilename(), ex);
             studentGroupBatchBuilder
-                    .digest("") //TODO: should digest be optional?
-                    .message(ex.getMessage())
+                    .digest(Hex.encodeHexString(messageDigest.digest()).toUpperCase())
+                    .message("Exception processing upload: " + ex.getMessage())
                     .status(ImportStatus.BAD_DATA);
-            sgbImport = repository.create(studentGroupBatchBuilder.build());
+            return repository.create(studentGroupBatchBuilder.build());
         }
-        return sgbImport;
     }
 
-    // TODO: add more validation ?
-    private void validate(final InputStream inputStream) {
-        try (final Reader reader = new InputStreamReader(inputStream, Charset.forName("utf8"))) {
-            final Iterator<CSVRecord> recordIterator = CSVFormat.RFC4180.withFirstRecordAsHeader().parse(reader).iterator();
-            if (!recordIterator.hasNext()) throw new IllegalArgumentException("Empty file");
+    private void archiveFile(final MultipartFile file,
+                             final String digest,
+                             final User user) throws IOException {
+        try (final InputStream fileStream = file.getInputStream()) {
+            final String location = new LocationStrategy.GroupUploadContentLocationStrategy().location(digest);
+
+            //Check if file already exists in ArchiveService
+            if (archiveService.exists(location)) {
+                logger.debug("Upload already exists in S3: {}", file.getOriginalFilename());
+                return;
+            }
```

Comment:
If the file already exists with the same hash, I don't re-upload it to S3.  I could see a user wanting to "reset" their group permissions by re-uploading the same file, but there's no reason to re-send the same bits to S3.